cmd/releasego/get-upstream-commit.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/microsoft/go-infra/executil"
+	"github.com/microsoft/go-infra/goversion"
+	"github.com/microsoft/go-infra/subcmd"
+)
+
+func init() {
+	subcommands = append(subcommands, subcmd.Option{
+		Name:    "get-upstream-commit",
+		Summary: "Get the upstream commit for a given release version. Poll if not available.",
+		Description: `
+
+If the given version is an ordinary release, wait for the upstream Git tag. If it's a FIPS release,
+wait until the boring releases file lists this release.
+
+If the version contains a revision, it is ignored. The microsoft/go repository uses revision numbers
+so it can release multiple times per upstream version. The build note must either be unspecified (to
+indicate an ordinary release) or "fips" (to indicate the release is based on the boring branch).
+`,
+		Handle: handleWaitUpstream,
+	})
+}
+
+func handleWaitUpstream(p subcmd.ParseFunc) error {
+	version := flag.String("version", "", "[Required] A full or partial microsoft/go version number (major.minor.patch[-revision[-suffix]]).")
+	upstream := flag.String("upstream", "https://go.googlesource.com/go", "The upstream Git repo to check.")
+	azdoVarName := flag.String("set-azdo-variable", "", "An AzDO variable name to set to the commit hash using a logging command.")
+	keepTemp := flag.Bool("w", false, "Keep the temporary repository used for polling, rather than cleaning it up.")
+	pollDelaySeconds := flag.Int("poll-delay", 5, "Number of seconds to wait between each poll attempt.")
```

Comment:
For a normal release, it should complete within a day. For a dev.boringcrypto branch release, maybe three days? This command doesn't actually have a timeout though, `-poll-delay` is just some time to wait between polls that we can ramp up if we hit a rate limit at some point. (A timeout will be enforced by AzDO rather than this command.)